File: backend/src/models/pattern.py
"""Pydantic schemas for Pattern Engine (Story 11.1).

Defines request/response schemas for pattern sessions and pieces.
"""

from decimal import Decimal
from datetime import datetime
from enum import Enum
from uuid import UUID

from typing import Optional

from pydantic import BaseModel, Field


class PatternSessionStatus(str, Enum):
    """Pattern session lifecycle statuses."""

    draft = "draft"
    completed = "completed"
    exported = "exported"


class PieceType(str, Enum):
    """Pattern piece types (collar added Story 11.8)."""

    front_bodice = "front_bodice"
    back_bodice = "back_bodice"
    sleeve = "sleeve"
    collar = "collar"


class SleeveType(str, Enum):
    """Sleeve construction style (FR91a, Story 11.7)."""

    raglan = "raglan"     # tay liền cổ — sleeve drafted to the neckline
    set_in = "set_in"     # tay tra — cap drafted from the body armhole (FR93)


class PatternSessionCreate(BaseModel):
    """Request schema for creating a pattern session.

    10 measurements are auto-filled from customer profile, editable by Owner.
    Validation ranges based on Vietnamese tailoring standards (FR99).
    """

    customer_id: UUID
    garment_type: str = Field(default="ao_dai", max_length=50)
    sleeve_type: SleeveType = Field(default=SleeveType.raglan, description="raglan | set_in (FR91a)")
    notes: str | None = Field(None, max_length=2000)

    # 10 body measurements (cm) with min/max validation
    do_dai_ao: Decimal = Field(..., ge=30, le=200, description="Do dai ao (cm)")
    ha_eo: Decimal = Field(..., ge=5, le=50, description="Ha eo (cm)")
    vong_co: Decimal = Field(..., ge=20, le=60, description="Vong co (cm)")
    vong_nach: Decimal = Field(..., ge=25, le=70, description="Vong nach (cm)")
    vong_nguc: Decimal = Field(..., ge=50, le=180, description="Vong nguc (cm)")
    vong_eo: Decimal = Field(..., ge=40, le=160, description="Vong eo (cm)")
    vong_mong: Decimal = Field(..., ge=50, le=180, description="Vong mong (cm)")
    do_dai_tay: Decimal = Field(..., ge=30, le=100, description="Do dai tay (cm)")
    vong_bap_tay: Decimal = Field(..., ge=15, le=60, description="Vong bap tay (cm)")
    vong_co_tay: Decimal = Field(..., ge=10, le=35, description="Vong co tay (cm)")

    # 5 extended measurements (Story 11.8) — optional; required by style (set-in: vai; darts: nguc)
    ha_ben_nguc: Decimal | None = Field(None, ge=15, le=35, description="Ha ben nguc (cm)")
    dang_nguc: Decimal | None = Field(None, ge=10, le=25, description="Dang nguc (cm)")
    ha_mong: Decimal | None = Field(None, ge=10, le=30, description="Ha mong — eo→mong (cm)")
    xuoi_vai: Decimal | None = Field(None, ge=1, le=8, description="Xuoi vai (cm)")
    rong_vai: Decimal | None = Field(None, ge=28, le=50, description="Rong vai (cm)")


class GeometryParams(BaseModel):
    """Structure for pattern_pieces.geometry_params JSONB.

    Computed geometric parameters from the deterministic formula engine.
    Corrected per SCP 2026-06-08 — front/back bodice now diverge in multiple
    dimensions; sleeve uses raglan drafting. Stored as a free dict in the DB, so
    this model documents the shape but is not strictly enforced on read.
    """

    model_config = {"extra": "allow"}

    piece: Optional[str] = None
    # Bodice-specific (None for sleeve pieces)
    length: Optional[float] = None
    armhole_drop: Optional[float] = None
    waist_drop: Optional[float] = None
    hip_drop: Optional[float] = None
    neck_width: Optional[float] = None
    neck_depth: Optional[float] = None
    shoulder_rise: Optional[float] = None
    bust_width: Optional[float] = None
    waist_width: Optional[float] = None
    hip_width: Optional[float] = None
    hem_width: Optional[float] = None
    seam_allowance: float = 1.0
    # Sleeve-specific (None for bodice pieces) — raglan
    sleeve_type: Optional[str] = None
    sleeve_length: Optional[float] = None
    bicep_offset: Optional[float] = None
    bicep_width: Optional[float] = None
    underarm_width: Optional[float] = None
    wrist_width: Optional[float] = None
    neck_front: Optional[float] = None
    neck_back: Optional[float] = None
    neck_rise: Optional[float] = None


class PatternPieceResponse(BaseModel):
    """Response schema for a single pattern piece."""

    id: UUID
    session_id: UUID
    piece_type: PieceType
    svg_data: str
    geometry_params: dict
    created_at: datetime

    model_config = {"from_attributes": True}


class PatternSessionResponse(BaseModel):
    """Response schema for a pattern session with pieces."""

    id: UUID
    tenant_id: UUID
    customer_id: UUID
    created_by: UUID
    do_dai_ao: Decimal
    ha_eo: Decimal
    vong_co: Decimal
    vong_nach: Decimal
    vong_nguc: Decimal
    vong_eo: Decimal
    vong_mong: Decimal
    do_dai_tay: Decimal
    vong_bap_tay: Decimal
    vong_co_tay: Decimal
    ha_ben_nguc: Decimal | None = None
    dang_nguc: Decimal | None = None
    ha_mong: Decimal | None = None
    xuoi_vai: Decimal | None = None
    rong_vai: Decimal | None = None
    garment_type: str
    sleeve_type: SleeveType = SleeveType.raglan  # default for pre-11.7 rows without the column
    notes: str | None
    status: PatternSessionStatus
    pieces: list[PatternPieceResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/pattern.ts
/**
 * Pattern Session Types and Zod Schemas
 * Story 11.4: Profile-Driven Measurement Form UI
 *
 * Defines types for pattern generation workflow:
 * - 10 measurement fields with Vietnamese labels
 * - Validation ranges for each measurement
 * - Pattern session create/response schemas
 */

import { z } from "zod";

// ===== Measurement Validation Ranges (AC #6) =====

export const MEASUREMENT_RANGES = {
  do_dai_ao: { min: 30, max: 200, label: "Độ dài áo" },
  ha_eo: { min: 5, max: 50, label: "Hạ eo" },
  vong_co: { min: 20, max: 60, label: "Vòng cổ" },
  vong_nach: { min: 30, max: 80, label: "Vòng nách" },
  vong_nguc: { min: 60, max: 150, label: "Vòng ngực" },
  vong_eo: { min: 50, max: 140, label: "Vòng eo" },
  vong_mong: { min: 60, max: 160, label: "Vòng mông" },
  do_dai_tay: { min: 30, max: 100, label: "Độ dài tay" },
  vong_bap_tay: { min: 15, max: 60, label: "Vòng bắp tay" },
  vong_co_tay: { min: 10, max: 40, label: "Vòng cổ tay" },
} as const;

export type MeasurementKey = keyof typeof MEASUREMENT_RANGES;

// ===== Sleeve type (FR91a, Story 11.7) =====
export type SleeveType = "raglan" | "set_in";

export const SLEEVE_TYPE_LABELS: Record<SleeveType, string> = {
  raglan: "Tay liền cổ (raglan)",
  set_in: "Tay tra (có đường may vai)",
};

// ===== Extended measurements (Story 11.8) — optional, by style =====
export const EXTENDED_MEASUREMENT_RANGES = {
  ha_ben_nguc: { min: 15, max: 35, label: "Hạ ben ngực" },
  dang_nguc: { min: 10, max: 25, label: "Dang ngực" },
  ha_mong: { min: 10, max: 30, label: "Hạ mông" },
  xuoi_vai: { min: 1, max: 8, label: "Xuôi vai" },
  rong_vai: { min: 28, max: 50, label: "Rộng vai" },
} as const;

export type ExtendedMeasurementKey = keyof typeof EXTENDED_MEASUREMENT_RANGES;

// ===== Pattern Measurement Fields Configuration (AC #4) =====

export const PATTERN_MEASUREMENT_FIELDS: Array<{
  key: MeasurementKey;
  label: string;
  description: string;
}> = [
  { key: "do_dai_ao", label: "Độ dài áo", description: "Body length (cm)" },
  { key: "ha_eo", label: "Hạ eo", description: "Waist drop (cm)" },
  { key: "vong_co", label: "Vòng cổ", description: "Neck circumference (cm)" },
  { key: "vong_nach", label: "Vòng nách", description: "Armhole circumference (cm)" },
  { key: "vong_nguc", label: "Vòng ngực", description: "Bust circumference (cm)" },
  { key: "vong_eo", label: "Vòng eo", description: "Waist circumference (cm)" },
  { key: "vong_mong", label: "Vòng mông", description: "Hip circumference (cm)" },
  { key: "do_dai_tay", label: "Độ dài tay", description: "Sleeve length (cm)" },
  { key: "vong_bap_tay", label: "Vòng bắp tay", description: "Bicep circumference (cm)" },
  { key: "vong_co_tay", label: "Vòng cổ tay", description: "Wrist circumference (cm)" },
];

// ===== Zod Schema for Measurement Validation =====

/**
 * Create a measurement field schema with Vietnamese error messages (AC #6)
 */
const createMeasurementField = (key: MeasurementKey) => {
  const { min, max, label } = MEASUREMENT_RANGES[key];
  return z
    .number({ message: `${label} phải là số` })
    .min(min, `${label} phải từ ${min} đến ${max} cm`)
    .max(max, `${label} phải từ ${min} đến ${max} cm`);
};

/**
 * Schema for 10 pattern measurement fields
 */
export const PatternMeasurementSchema = z.object({
  do_dai_ao: createMeasurementField("do_dai_ao"),
  ha_eo: createMeasurementField("ha_eo"),
  vong_co: createMeasurementField("vong_co"),
  vong_nach: createMeasurementField("vong_nach"),
  vong_nguc: createMeasurementField("vong_nguc"),
  vong_eo: createMeasurementField("vong_eo"),
  vong_mong: createMeasurementField("vong_mong"),
  do_dai_tay: createMeasurementField("do_dai_tay"),
  vong_bap_tay: createMeasurementField("vong_bap_tay"),
  vong_co_tay: createMeasurementField("vong_co_tay"),
});

export type PatternMeasurementInput = z.infer<typeof PatternMeasurementSchema>;

// ===== Pattern Session Create Schema (AC #7) =====

const optionalExtended = (key: ExtendedMeasurementKey) => {
  const { min, max, label } = EXTENDED_MEASUREMENT_RANGES[key];
  return z
    .number()
    .min(min, `${label} phải từ ${min} đến ${max} cm`)
    .max(max, `${label} phải từ ${min} đến ${max} cm`)
    .optional();
};

export const PatternSessionCreateSchema = z.object({
  customer_id: z.string().uuid().nullable(),
  garment_type: z.string().default("ao_dai"),
  sleeve_type: z.enum(["raglan", "set_in"]).default("raglan"),
  notes: z.string().optional(),
  // Spread the 10 required measurement fields
  ...PatternMeasurementSchema.shape,
  // 5 optional extended measurements (Story 11.8)
  ha_ben_nguc: optionalExtended("ha_ben_nguc"),
  dang_nguc: optionalExtended("dang_nguc"),
  ha_mong: optionalExtended("ha_mong"),
  xuoi_vai: optionalExtended("xuoi_vai"),
  rong_vai: optionalExtended("rong_vai"),
});

export type PatternSessionCreate = z.infer<typeof PatternSessionCreateSchema>;

// ===== Pattern Pieces =====

export type PieceType = "front_bodice" | "back_bodice" | "sleeve" | "collar";

export type ExportFormat = "svg" | "gcode";

export interface PatternPieceResponse {
  id: string;
  session_id: string;
  piece_type: PieceType;
  svg_data: string | null;
  geometry_params: Record<string, unknown> | null;
  created_at: string;
}

// ===== Pattern Session Response Types =====

export interface PatternSessionResponse {
  id: string;
  tenant_id: string;
  customer_id: string | null;
  created_by: string;
  garment_type: string;
  sleeve_type: SleeveType;
  status: "draft" | "completed" | "exported";
  do_dai_ao: number;
  ha_eo: number;
  vong_co: number;
  vong_nach: number;
  vong_nguc: number;
  vong_eo: number;
  vong_mong: number;
  do_dai_tay: number;
  vong_bap_tay: number;
  vong_co_tay: number;
  ha_ben_nguc: number | null;
  dang_nguc: number | null;
  ha_mong: number | null;
  xuoi_vai: number | null;
  rong_vai: number | null;
  notes: string | null;
  pieces: PatternPieceResponse[];
  created_at: string;
  updated_at: string;
}

// ===== Customer Measurement Mapping =====

/**
 * Map customer measurement fields to pattern session fields
 * Some pattern fields don't have customer measurement equivalents
 */
export const CUSTOMER_TO_PATTERN_MAPPING: Record<string, MeasurementKey | null> = {
  top_length: "do_dai_ao",
  ha_eo: "ha_eo",
  neck: "vong_co",
  vong_nach: "vong_nach",
  bust: "vong_nguc",
  waist: "vong_eo",
  hip: "vong_mong",
  sleeve_length: "do_dai_tay",
  vong_bap_tay: "vong_bap_tay",
  wrist: "vong_co_tay",
};

/**
 * Reverse mapping: pattern field → customer measurement field
 */
export const PATTERN_TO_CUSTOMER_MAPPING: Record<MeasurementKey, string | null> = {
  do_dai_ao: "top_length",
  ha_eo: "ha_eo",
  vong_co: "neck",
  vong_nach: "vong_nach",
  vong_nguc: "bust",
  vong_eo: "waist",
  vong_mong: "hip",
  do_dai_tay: "sleeve_length",
  vong_bap_tay: "vong_bap_tay",
  vong_co_tay: "wrist",
};

export interface PatternSessionListItem {
  id: string;
  status: "draft" | "completed" | "exported";
  garment_type: string;
  piece_count: number;
  created_at: string;
}

export const PIECE_TYPE_LABELS: Record<PieceType, string> = {
  front_bodice: "Thân trước",
  back_bodice: "Thân sau",
  sleeve: "Tay áo",
  collar: "Lá cổ",
};

export const PATTERN_SESSION_STATUS_LABELS: Record<string, string> = {
  draft: "Nháp",
  completed: "Hoàn thành",
  exported: "Đã xuất",
};

export const GEOMETRY_PARAM_LABELS: Record<string, string> = {
  length: "Dài",
  bust_width: "Rộng ngực",
  waist_width: "Rộng eo",
  hip_width: "Rộng mông",
  hem_width: "Rộng tà",
  neck_width: "Rộng cổ",
  neck_depth: "Sâu cổ",
  armhole_drop: "Hạ nách",
  sleeve_length: "Dài tay",
  bicep_width: "Rộng bắp tay",
  underarm_width: "Rộng nách tay",
  wrist_width: "Cửa tay",
  cap_height: "Cao đầu tay",
  cap_ease: "Cử động đầu tay",
  band: "Bản cổ",
  neck_perimeter: "Chu vi cổ",
};

// ===== Story 11.6: Attach Pattern Types =====

export interface AttachPatternRequest {
  pattern_session_id: string | null;
}

export interface AttachPatternResponse {
  order_id: string;
  pattern_session_id: string | null;
}

// Keys aligned with the actual engine output (SCP 2026-06-08). The sleeve list covers
// both raglan and set-in fields; only the present ones render.
export const PIECE_GEOMETRY_PARAM_KEYS: Record<PieceType, string[]> = {
  front_bodice: ["length", "bust_width", "waist_width", "hip_width", "hem_width", "neck_width", "neck_depth", "armhole_drop"],
  back_bodice: ["length", "bust_width", "waist_width", "hip_width", "hem_width", "neck_width", "armhole_drop"],
  sleeve: ["sleeve_length", "bicep_width", "underarm_width", "wrist_width", "armhole_drop", "cap_height", "cap_ease"],
  collar: ["length", "band", "neck_perimeter"],
};
